metron-platform/metron-writer/src/main/java/org/apache/metron/writer/BatchTimeoutPolicy.java

```diff
@@ -0,0 +1,104 @@
+  //This is a conservative defaultBatchTimeout for a vanilla bolt with batchTimeoutDivisor=2
+  public static final int UNINITIALIZED_DEFAULT_BATCH_TIMEOUT = 6;
+  private int defaultBatchTimeout = UNINITIALIZED_DEFAULT_BATCH_TIMEOUT;
+  private Clock clock = new Clock();
+  private Map<String, Long> timeouts = new HashMap<>();
+
+  public void setClock(Clock clock) {
+    this.clock = clock;
+  }
+
+  public void setDefaultBatchTimeout(int defaultBatchTimeout) {
+    this.defaultBatchTimeout = defaultBatchTimeout;
+  }
+
+  /**
+   * Manages timeouts for each sensor type and determines when a batch should be flushed.  At the start of a new
+   * batch the timeout value is computed based on the current time and configured timeout value.  Subsequent calls check to
+   * see if the timeout has been reached and flushes if so.  A reset clears the timeout value for that sensor type.
+   * @param sensorType sensor type
+   * @param configurations writer configurations includes timeouts
+   * @param batchSize number of messages to be written (not used here)
+   * @return true if the timeout has been reached
+   */
+  @Override
+  public boolean shouldFlush(String sensorType, WriterConfiguration configurations, int batchSize) {
+    boolean shouldFlush = false;
+    long currentTimeMillis = clock.currentTimeMillis();
+    if (!timeouts.containsKey(sensorType)) {  // no timeout present so assume this is a new batch
+      //This block executes at the beginning of every batch, per sensor.
+      //configurations can change, so (re)init getBatchTimeout(sensorType) at start of every batch
+      long batchTimeoutMs = getBatchTimeout(sensorType, configurations);
+      LOG.debug("Setting batch timeout to {} for sensor {}.", batchTimeoutMs, sensorType);
+      timeouts.put(sensorType, currentTimeMillis + batchTimeoutMs);
+    }
+    if (timeouts.get(sensorType) <= currentTimeMillis) {
+      LOG.debug("Batch timeout of {} reached. Flushing {} messages for sensor {}.",
+              timeouts.get(sensorType), batchSize, sensorType);
+      shouldFlush = true;
+    }
+    return shouldFlush;
+  }
+
+  /**
+   * Removes the timeout value for a sensor type.  The next call to {@link org.apache.metron.writer.BatchTimeoutPolicy#shouldFlush(String, WriterConfiguration, int)}
+   * will set a new timeout.
+   * @param sensorType
+   */
+  @Override
+  public void reset(String sensorType) {
+    timeouts.remove(sensorType);
+  }
+
+  /**
+   * Returns the configured timeout for a sensor type in milliseconds.  The default timeout will be used if the configured timeout is
+   * set to 0 or greater than the default timeout.
+   * @param sensorType
+   * @param configurations
+   * @return
+   */
+  protected long getBatchTimeout(String sensorType, WriterConfiguration configurations) {
+    int batchTimeoutSecs = configurations.getBatchTimeout(sensorType);
+    if (batchTimeoutSecs <= 0 || batchTimeoutSecs > defaultBatchTimeout) {
```

Comment:
This was preexisting.  I don't know the exact reason but I suspect it was a guard against setting the batch timeout to be greater than the tuple timeout.  @mmiklavc also brought up a question around default batch timeout and I'm not quite sure if what we have is correct or what the requirements are.  I know that the default batch timeout is 1/2 the tuple timeout when running in a bolt and 6 otherwise (test scenarios I think).  I also know the default batch timeout is only set once at the beginning and isn't tied to anything dynamically. 

How should default batch timeout work?  I think we need to keep the default to be 1/2 the tuple time in storm but what about other scenarios?  Should this setting be exposed in say global config?  Should we be able to change it at runtime?

Reply:
I think we should just do what the user sets it to here.  Changing it under the covers like this is just going to cause confusion.  Especially since this code will also be run outside of Storm, it shouldn't be bound by subtle Storm constraints.

Perhaps there is somewhere in Storm-specific code that we can add a check and clear log statement or exception, if the configured batch timeout exceeds some percentage of the tuple timeout?  `ParserBolt` maybe or elsewhere?

If this is current behavior, I'd also be open to just tabling this for a separate JIRA/PR.

Reply:
Hrm, that's a good catch @nickwallen. And I believe @merrimanr is right. The reason we want to restrict the timeout range is so that if the tuple timeout ends up being short enough with respect to our batch timeout, we'll fail a lot of tuples due to timeout. I'm not sure if there's a better way to capture that reasoning in the code, but we should definitely refactor or add comments to make this clearer.

It seems likely that the only reason this was configurable (the default, that is) in the first place was for testing. And oddly enough, that capability and flexibility made this less clear for us now. I think ultimately in Storm world we should have that default lockstep with the Storm concept of tuple timeout. So if it's 1/2 the tuple timeout, let's leave that be. Considering other streaming frameworks, that concept of where we derive the default should be more general. 

I would be in favor of renaming "defaultBatchTimeout" to "maxBatchTimeout" and then having a method "setMaxBatchTimeout" or constructor arg that clearly sets that value. Any reliance on tuples or Storm-specific config I think should be kept out of the timeout policy implementation and provided externally. In this case, this is an option that is not dynamic bc changing the tuple timeout requires a topology restart. We should continue to pull the current tuple timeout from the Storm Bolt config and pass that in per the calculation as the "max timeout." Our Storm code cares about the max being 1/2 tuple timeout whereas the policy class only cares that there's the concept of a max value in the first place. i.e. let's leave the construction of that value to the classes that are instantiating the policies. How's that sound to you gents?

Reply:
The latest commit changes "defaultBatchTimeout" to "maxBatchTimeout", moves that setting to the constructor(s) and removes the set methods.  A default value no longer lives in `BatchTimeoutPolicy` and instead an exception is thrown if the value is less than or equal to 0.  The bolts now pass down the value into the `BulkWriterComponent` constructor which uses it to create the policy.

There was a default value of 6 being used previously.  As I was making these changes I realized the `WriterBolt` class relies on this default instead of calculating what it should be.  I moved the default to this class instead so it continues to function as it does now.  Rather than going to the trouble of figuring out the right way to set this I think we should replace it with the `BulkMessageWriterBolt`.  They do the same thing and are redundant.  Planning on opening a separate PR for that anyways.  If anyone disagrees with this and would like to discuss the correct way to set "maxBatchTimeout" in this bolt let me know.